core/database_util.go

```diff
@@ -276,22 +325,36 @@ func WriteHeader(db ethdb.Database, header *types.Header) error {
 	if err != nil {
 		return err
 	}
-	key := append(append(blockPrefix, header.Hash().Bytes()...), headerSuffix...)
+	hash := header.Hash().Bytes()
+	num := header.Number.Uint64()
+	encNum := encodeBlockNumber(num)
+	key := append(blockHashPrefix, hash...)
+	if err := db.Put(key, encNum); err != nil {
+		glog.Fatalf("failed to store hash to number mapping into database: %v", err)
+		return err
+	}
+	if num == 0 {
+		if err := db.Put([]byte("useSequentialKeys"), []byte{1}); err != nil {
+			glog.Fatalf("failed to store sequential key flag: %v", err)
+			return err
+		}
+	}
```

Comment:
As agreed, this upgrade mechanism key should be handled in the upgrade mechanism, not here.